src/NzbDrone.Core/Download/Clients/DownloadStation/SharedFolderResolverService.cs

```diff
@@ -0,0 +1,45 @@
+﻿using NLog;
+using NzbDrone.Common.Cache;
+using NzbDrone.Common.Disk;
+using NzbDrone.Core.Configuration;
+using NzbDrone.Core.Download.Clients.DownloadStation.Proxies;
+using NzbDrone.Core.ThingiProvider;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace NzbDrone.Core.Download.Clients.DownloadStation
+{
+    public class SharedFolderResolverService : ISharedFolderResolverService
+    {
+        private readonly IFileStationProxy _proxy;
+        private ICached<SharedFolderCache> _cache;
+        private readonly ILogger _logger;
+
+        public DownloadStationSettings Settings { get; set; }
+
+        public SharedFolderResolverService(ICacheManager cacheManager,
+                                    IFileStationProxy proxy,
+                                    Logger logger)
+        {
+            _proxy = proxy;
+            _cache = cacheManager.GetCache<SharedFolderCache>(GetType());
+            _logger = logger;
+        }
+
+        private string GetPhysicalPath(string sharedFolder)
+        {
+            var physicalPath = _proxy.GetPhysicalPath(sharedFolder, Settings);
+            var sharedFolderCache = new SharedFolderCache(physicalPath);
+            _cache.Set(sharedFolder, sharedFolderCache);
+
+            return physicalPath;
+        }
+
+        public OsPath ResolvePhysicalPath(string sharedFolder)
+        {
+            return new OsPath(_cache.Find(sharedFolder)?.PhysicalPath ?? GetPhysicalPath(sharedFolder));
```

Comment:
ICached has a nice [method](https://github.com/Sonarr/Sonarr/blob/816cf608fcef07e375aa750a863d0a7365108981/src/NzbDrone.Common/Cache/ICached.cs#L17), including the lifetime property.
So `_cache.Get(sharedFolder, () => GetPhysicalPath(sharedFolder), TimeSpan.FromHour(1));` then you remove the _cache.Set from GetPhysicalPath, and everything sorts itself out and expires every once in a while. It's also more atomic.

You asked about the Cache lifetime. Just pick something sensible. It's fairly arbitrary, but 1h seemed reasonable, but feel free to increase it to 3h or so.